File: backend/app/thinning.py
"""Viewport thinning (§5): cap the displayed set, spread it geographically
with a grid (best per cell first), keep denylisted results visible."""
import math


def quality_key(r: dict) -> tuple:
    """Rating-weighted ranking proxy used to pick within cells.

    Scores aren't computed yet at thinning time (the location-count lookup
    only runs for displayed candidates, §6), so rating x log(reviews) stands
    in. Denylist status is already known and handled by the caller.
    """
    rating = r.get("rating") or 0
    reviews = r.get("userRatingCount") or 0
    return (rating * math.log10(reviews + 1), reviews)


def thin_results(results: list[dict], bounds: dict, cap: int, grid: int) -> list[dict]:
    """results: candidate dicts with lat/lng/rating/userRatingCount/denylisted."""
    denylisted = [r for r in results if r.get("denylisted")]
    candidates = [r for r in results if not r.get("denylisted")]
    slots = max(0, cap - len(denylisted))
    if len(candidates) <= slots:
        return candidates + denylisted

    lat_span = (bounds["north"] - bounds["south"]) or 1e-9
    lng_span = (bounds["east"] - bounds["west"]) or 1e-9

    def cell(r: dict) -> tuple[int, int]:
        row = min(grid - 1, max(0, int((r["lat"] - bounds["south"]) / lat_span * grid)))
        col = min(grid - 1, max(0, int((r["lng"] - bounds["west"]) / lng_span * grid)))
        return (row, col)

    by_cell: dict[tuple[int, int], list[dict]] = {}
    for r in candidates:
        by_cell.setdefault(cell(r), []).append(r)
    for members in by_cell.values():
        members.sort(key=quality_key, reverse=True)

    # Round-robin: take the best from every cell, then second-best, etc.,
    # so results spread across the viewport before any cell gets a second pick.
    kept: list[dict] = []
    depth = 0
    while len(kept) < slots:
        layer = [m[depth] for m in by_cell.values() if depth < len(m)]
        if not layer:
            break
        layer.sort(key=quality_key, reverse=True)
        kept.extend(layer[: slots - len(kept)])
        depth += 1

    return kept + denylisted

File: worker/thinning.ts
import type { Bounds, Candidate } from "./types";

function quality(candidate: Candidate): [number, number] {
  return [(candidate.rating ?? 0) * Math.log10(candidate.userRatingCount + 1), candidate.userRatingCount];
}

function sortByQuality(candidates: Candidate[]): void {
  candidates.sort((left, right) => {
    const a = quality(left);
    const b = quality(right);
    return b[0] - a[0] || b[1] - a[1];
  });
}

export function thinResults(results: Candidate[], bounds: Bounds, cap: number, grid: number): Candidate[] {
  const denylisted = results.filter((result) => result.denylisted);
  const candidates = results.filter((result) => !result.denylisted);
  const slots = Math.max(0, cap - denylisted.length);
  if (candidates.length <= slots) return [...candidates, ...denylisted];

  const latSpan = bounds.north - bounds.south || 1e-9;
  const lngSpan = bounds.east - bounds.west || 1e-9;
  const cells = new Map<string, Candidate[]>();
  for (const candidate of candidates) {
    const row = Math.min(
      grid - 1,
      Math.max(0, Math.floor(((candidate.lat - bounds.south) / latSpan) * grid))
    );
    const column = Math.min(
      grid - 1,
      Math.max(0, Math.floor(((candidate.lng - bounds.west) / lngSpan) * grid))
    );
    const key = `${row},${column}`;
    const cell = cells.get(key) ?? [];
    cell.push(candidate);
    cells.set(key, cell);
  }
  for (const cell of cells.values()) sortByQuality(cell);

  const kept: Candidate[] = [];
  for (let depth = 0; kept.length < slots; depth += 1) {
    const layer = [...cells.values()].flatMap((cell) => (cell[depth] ? [cell[depth]] : []));
    if (layer.length === 0) break;
    sortByQuality(layer);
    kept.push(...layer.slice(0, slots - kept.length));
  }
  return [...kept, ...denylisted];
}
